backends-velox/src/main/scala/org/apache/spark/sql/execution/unsafe/UnsafeColumnarBuildSideRelation.scala

```diff
@@ -0,0 +1,312 @@
+
+/**
+ * UnsafeColumnarBuildSideRelation should backed by offheap to avoid on-heap oom. Almost the same as
+ * ColumnarBuildSideRelation, we should remove ColumnarBuildSideRelation when
+ * UnsafeColumnarBuildSideRelation get matured.
+ *
+ * @param output
+ * @param batches
+ */
+case class UnsafeColumnarBuildSideRelation(
+    private var output: Seq[Attribute],
+    private var batches: UnsafeBytesBufferArray,
+    var mode: BroadcastMode)
+  extends BuildSideRelation
+  with Externalizable
+  with Logging
+  with KryoSerializable {
+
+  // Needed for serialization
+  def this() = {
+    this(null, null.asInstanceOf[UnsafeBytesBufferArray], null)
+  }
+
+  def this(output: Seq[Attribute], bytesBufferArray: Array[Array[Byte]], mode: BroadcastMode) = {
+    // only used in driver side when broadcast the whole batches
+    this(
+      output,
+      UnsafeBytesBufferArray(
+        bytesBufferArray.length,
+        bytesBufferArray.map(_.length),
+        bytesBufferArray.map(_.length.toLong).sum,
+        TaskContext.get().taskMemoryManager
+      ),
+      mode
+    )
+    val batchesSize = bytesBufferArray.length
+    for (i <- 0 until batchesSize) {
+      val length = bytesBufferArray(i).length
+      log.debug(s"this $i--- $length")
+      batches.putBytesBuffer(i, bytesBufferArray(i))
+    }
+  }
+
+  // should only be used on driver to serialize this relation
+  override def writeExternal(out: ObjectOutput): Unit = Utils.tryOrIOException {
+    out.writeObject(output)
+    out.writeObject(mode)
+    out.writeInt(batches.arraySize)
+    out.writeObject(batches.bytesBufferLengths)
+    out.writeLong(batches.totalBytes)
+    for (i <- 0 until batches.arraySize) {
+      val bytes = batches.getBytesBuffer(i)
+      out.write(bytes)
+      log.debug(s"writeExternal index $i with length ${bytes.length}")
+    }
+  }
+
+  // should only be used on driver to serialize this relation
+  override def write(kryo: Kryo, out: Output): Unit = Utils.tryOrIOException {
+    kryo.writeObject(out, output.toList)
+    kryo.writeObject(out, mode)
+    out.writeInt(batches.arraySize)
+    kryo.writeObject(out, batches.bytesBufferLengths)
+    out.writeLong(batches.totalBytes)
+    for (i <- 0 until batches.arraySize) {
+      val bytes = batches.getBytesBuffer(i)
+      out.write(bytes)
+      log.debug(s"write index $i with length ${bytes.length}")
+    }
+  }
+
+  // should only be used on executor to deserialize this relation
+  override def readExternal(in: ObjectInput): Unit = Utils.tryOrIOException {
+    output = in.readObject().asInstanceOf[Seq[Attribute]]
+    mode = in.readObject().asInstanceOf[BroadcastMode]
+    val totalArraySize = in.readInt()
+    val bytesBufferLengths = in.readObject().asInstanceOf[Array[Int]]
+    val totalBytes = in.readLong()
+
+    val taskMemoryManager = new TaskMemoryManager(
+      new UnifiedMemoryManager(SparkEnv.get.conf, Long.MaxValue, Long.MaxValue / 2, 1),
+      0)
```

Comment:
Would you like to explain a bit against this code? Thanks!

Reply:
This is used in Broadcast, shared by multiple tasks, so here introduce a new TaskMemoryManager instead of use task-related TaskMemoryManager.
Similar to `UnsafeHashRelation` in https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashedRelation.scala#L389-L410

Reply:
Thanks for the explanation. And I need to apologize that my previous comment https://github.com/apache/incubator-gluten/issues/7750#issuecomment-2451909071 was wrong, because we are apparently facing the similar issue with Gazelle project here again.

I see `finalize()` is used and I assume it's working in your test, right? And given that invocations on `finalize()` relies on JVM GC, how do we make sure the off-heap memory can be released to avoid off-heap overhead, even JVM GC is never triggered?

This code (line :127 - :129) looks feasible but the allocation is still not managed by the default memory manager. Do we have a better approach? Otherwise [Yarn kill](https://github.com/apache/incubator-gluten/issues/6947) issue will be led.

We can revisit the topic and the code a little bit and hopefully we can come up with an optimal solution. My description in https://github.com/apache/incubator-gluten/issues/7750 was indeed a bit misleading, this issue doesn't look to have some kind of `simple` solution so far.

Reply:
@zhztheplayer Spark uses the GC mechanism of Cleaner to clean up broadcasts. `finalize()` should also be a feasible solution, but it cannot be actively cleaned up when offheap is insufficient. As for the need for a unified memory manager to manage this memory, we can directly use SparkEnv.get.memoryManager. see https://github.com/apache/incubator-gluten/pull/8349

Reply:
Edit: 

Broadcast variables can usually be assumed to be small enough, and the way of `finalize()` is roughly acceptable (maybe I'm wrong). Later, we can look for a more suitable solution, such as implementing the spill of UnsafeBytesBufferArray?

Reply:
@zhztheplayer Hi, terry is on vacation recently. What's your opinion on this idea? I can continue to follow up.

Reply:
@yikf Thank you for continuing the work.  I was on vacation either today so sorry for the late reply.

> @zhztheplayer Hi, terry is on vacation recently. What's your opinion on this idea? I can continue to follow up.

I was thinking about using storage memory but the implementation is complicated? If so let's put that idea on hold.

> we can directly use SparkEnv.get.memoryManager.

👍 Sounds good to me if it can utilize Spark's off-heap memory, since at least the memory is tracked. Though the release could still be a problem? Because Spark doesn't do `System.gc()` on off-heap OOM. 

If the issue does apply, maybe we can consider one of the following:

Option 1. Hack into Spark to make sure `System.gc()` is triggered on OOM. (if there's an apporach without modifications on vanilla Spark's code)
Option 2. Set a fixed capacity (e.g., 15% of off-heap memory) for the outstanding task memory manager used in the PR. When this part of memory is run out, trigger `System.gc()`, then if still unsatisified, throw an OOM.

The above are just based on my assumption. Let me know if any other possible solutions.

Reply:
@zhztheplayer There is another idea whose feasibility is uncertain. 

Spark uses the GC mechanism on the Driver side to [clean broadcast variables](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ContextCleaner.scala#L205)through ContextCleaner and supports the [attachListener](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/ContextCleaner.scala#L123). We can fully utilize this mechanism.

Prerequisite,
1. Implement our own set of endpoints for the off-heap broadcast cleanup.
2. Setup the endpoint if use off-heap broadcast.
3. `UnsafeColumnarBuildSideRelation` record the broadcast id.
4. Implement ContextCleaner listener.
5. deserialize the `UnsafeColumnarBuildSideRelation` at executor side, use a singleton class record broadcast id and the `UnsafeColumnarBuildSideRelation` (read external). 

Then,
When the Driver triggers broadcast clearing, the listener catchs the event and sends the clearing request to the executor side using endpoint. The endpoint of the executor uses broadcastId to find the corresponding relation. Actively penalizes the release method to free offheap memory. don't use `finalize()`.

The idea is similar to cleaning the broadcast block on the executor side.

Not sure if it will work, but if it does, this idea seems to be a final solution that fits the current Spark mechanics.

Reply:
Thank you for the thoughtful insights.

More or less, the idea looks very to the solution CH backend is using. In case you happend to miss that one, could refer to code starting from [here](https://github.com/apache/incubator-gluten/blob/main/backends-clickhouse/src/main/scala/org/apache/spark/rpc/GlutenDriverEndpoint.scala).

Looks like feasible to me though there are 2 limitations coming out from my mind with the solution:

1. The GC timing of driver-side variable is still uncertain
2. What if the driver-side variable is GC-ed, while executor-side variable is still in use? (Perhaps, set driver-side variable to null manually, immediately after broadcasted to testify)

Reply:
#2 does not seem to happen, at least until the end of the current stage, functions on the driver side will refer to broadcast variables, not sure if this is the case.

The gc timing on the driver side is also not controllable, but let's revisit the topic, we need to free up offheap memory when the broadcast variable is not in use, but there doesn't seem to be a certain thing on the executor side to recognize that the broadcast is no longer in use, so we need to trigger it on the executor side via gc, which is also somewhat strange in nature... Whether it's actively triggered gc or triggered by the jvm, and active gc is actually not a good solution in my opinion, it affects a lot of things.

Reply:
> I was thinking about using storage memory but the implementation is complicated? If so let's put that idea on hold.
Using storage memory is a bit more complex and Spark broadcast mechanism basic implements is based on Spark's storage memory in nature. We can try this way later to see if things is ok.